### PR TITLE
UCP/CONTEXT: add ERROR_HANDLER_DELAY configuration

### PR DESCRIPTION
**Pull request by evgeny-leksikov**
## What
 - add ERROR_HANDLER_DELAY configuration
 - add test_ucp_sockaddr.close_ep_force_before_err_cb
 - related fixes in UCP and UCT
 - updated version of  of https://github.com/openucx/ucx/pull/5931

## Why
to handle a case then err handling callback is scheduled on slow path progress but EP was closed by ucp_ep_close_nbx(force) earlier

### CODE REVIEW

**Review by dmitrygx — APPROVED** [review]

File: src/ucp/core/ucp_context.c [hunk 1]
```diff
@@ -329,6 +329,12 @@ static ucs_config_field_t ucp_config_table[] = {
    "endpoint.",
    ucs_offsetof(ucp_config_t, ctx.proto_indirect_id), UCS_CONFIG_TYPE_ON_OFF_AUTO},
 
+  {"ERROR_HANDLER_DELAY", "0us",
+   "Time period between error has been detected and handled (0 - disabled).\n"
```
Comment (yosefe) [inline src/ucp/core/ucp_context.c:333]: Artificial delay between detecting an error and processing it (0 - disabled). Used for testing purposes.

File: src/ucp/core/ucp_worker.c [hunk 2]
```diff
@@ -432,6 +432,19 @@ static unsigned ucp_worker_iface_err_handle_progress(void *arg)
     ucs_status_t status                         = err_handle_arg->status;
     ucp_worker_h worker                         = ucp_ep->worker;
     ucp_request_t *close_req;
+    uct_worker_cb_id_t prog_id;
+
+    if (ucs_get_time() < err_handle_arg->timeout) {
+        ucs_debug("ep %p: defer err handler, flags: 0x%x, status %s",
```
Comment (yosefe) [inline src/ucp/core/ucp_worker.c:438]: `delay error handler by %.2f usec`

File: src/ucp/core/ucp_worker.c [hunk 3]
```diff
@@ -3018,8 +3033,6 @@ void ucp_worker_keepalive_remove_ep(ucp_ep_h ep)
         return;
     }
 
-    ucs_assert(!ucs_list_is_empty(&worker->all_eps));
```
Comment (yosefe) [inline src/ucp/core/ucp_worker.c:3021]: why removed?
if ep is on the list IMO it should be non-empty?

File: src/ucp/stream/stream_recv.c [hunk 4]
```diff
@@ -536,14 +536,6 @@ ucp_stream_am_handler(void *am_arg, void *am_data, size_t am_length,
     UCP_WORKER_GET_VALID_EP_BY_ID(&ep, worker, data->hdr.ep_id, return UCS_OK,
                                   "stream data");
     ep_ext = ucp_ep_ext_proto(ep);
-
-    if (ucs_unlikely(ep->flags & (UCP_EP_FLAG_CLOSED |
```
Comment (yosefe) [inline src/ucp/stream/stream_recv.c:540]: removed because it's handled by ep_id?
maybe add comment here?

File: src/uct/tcp/tcp_ep.c [hunk 5]
```diff
@@ -1058,6 +1058,7 @@ ucs_status_t uct_tcp_ep_handle_io_err(uct_tcp_ep_t *ep, const char *op_str,
 {
     uct_tcp_iface_t *iface = ucs_derived_of(ep->super.super.iface,
                                             uct_tcp_iface_t);
+    ucs_log_level_t log_level = UCS_LOG_LEVEL_DIAG;
```
Comment (yosefe) [inline src/uct/tcp/tcp_ep.c:1061]: align on column

File: src/uct/tcp/tcp_ep.c [hunk 6]
```diff
@@ -1068,8 +1069,11 @@ ucs_status_t uct_tcp_ep_handle_io_err(uct_tcp_ep_t *ep, const char *op_str,
         return UCS_ERR_NO_PROGRESS;
     }
 
-    if (!uct_tcp_ep_is_conn_closed_by_peer(io_status)) {
+    if (uct_tcp_ep_is_conn_closed_by_peer(io_status)) {
+        log_level = UCS_LOG_LEVEL_DIAG;
```
Comment (yosefe) [inline src/uct/tcp/tcp_ep.c:1073]: isn't it already set to diag?

File: test/gtest/ucp/test_ucp_sockaddr.cc [hunk 7]
```diff
@@ -66,6 +66,15 @@ class test_ucp_sockaddr : public ucp_test {
...
         modify_config("KEEPALIVE_INTERVAL", "10s");
         modify_config("CM_USE_ALL_DEVICES", cm_use_all_devices() ? "y" : "n");
+
+        /*
+         * FIXME: this is a workaround of the issue reproduced by
+         *        the 'close_ep_force_before_err_cb' test with RC transport
+         *        where TX-queue less than FC_WND, so uct_flush cancel returns
+         *        UCS_ERR_NO_RESOURCES and can not be handled after error
+         */
+        modify_config("RC_FC_WND_SIZE", "128", SETENV_IF_NOT_EXIST);
```
Comment (yosefe) [inline test/gtest/ucp/test_ucp_sockaddr.cc:76]: @dmitrygx can it be related to MTT issue with FC?

File: test/gtest/ucp/test_ucp_sockaddr.cc [hunk 8]
```diff
@@ -681,6 +691,31 @@ class test_ucp_sockaddr : public ucp_test {
 
 unsigned test_ucp_sockaddr::m_err_count = 0;
 
+UCS_TEST_P(test_ucp_sockaddr, close_ep_force_before_err_cb,
+           "ERROR_HANDLER_DELAY=1s") {
+    const size_t N = 3000;
```
Comment (yosefe) [inline test/gtest/ucp/test_ucp_sockaddr.cc:696]: pls rename N to something better

File: test/gtest/ucp/ucp_test.cc [hunk 9]
```diff
@@ -228,6 +228,8 @@ ucs_status_t ucp_test::request_process(void *req, int worker_index, bool wait)
             progress(worker_index);
             status = ucp_request_check_status(req);
         } while ((status == UCS_INPROGRESS) && (ucs_get_time() < deadline));
+        EXPECT_NE(UCS_INPROGRESS, status) << "request " << req
+                                          << " was not completed on time";
```
Comment (yosefe) [inline test/gtest/ucp/ucp_test.cc:232]: duplication of error in line 239?

File: src/ucp/core/ucp_worker.c (hunk 3 above)
Comment (evgeny-leksikov) [reply src/ucp/core/ucp_worker.c:3021]: new test produces the case when `ucp_worker_keepalive_remove_ep` can be called twice for the same EP - from ep_close(force) and later, when err handling is processed and EP discarded (ref count == 0), as discussed with @hoopoepg this function only checks if the iterator does not point to  the ep and advances if it does. so the assert is redundant

Comment (dmitrygx) [reply src/ucp/core/ucp_worker.c:3021]: yes, I agree that EP should be on the list when we removing it from KA

File: test/gtest/ucp/test_ucp_sockaddr.cc (hunk 7 above)
Comment (dmitrygx) [reply test/gtest/ucp/test_ucp_sockaddr.cc:76]: but we don't do flush-cancel there

File: test/gtest/ucp/ucp_test.cc (hunk 9 above)
Comment (evgeny-leksikov) [reply test/gtest/ucp/ucp_test.cc:232]: 👍 probably rebase issue, I'd remove that one, here it looks simpler

File: test/gtest/ucp/test_ucp_sockaddr.cc (hunk 7 above)
Comment (evgeny-leksikov) [reply test/gtest/ucp/test_ucp_sockaddr.cc:76]: ucp_ep_close(force) does

File: src/ucp/core/ucp_worker.c (hunk 3 above)
Comment (evgeny-leksikov) [reply src/ucp/core/ucp_worker.c:3021]: I guess currently we don't have a way to know if the ep on the list or not

File: test/gtest/ucp/test_ucp_sockaddr.cc (hunk 7 above)
Comment (dmitrygx) [reply test/gtest/ucp/test_ucp_sockaddr.cc:76]: > ucp_ep_close(force) does

yes, but we don't receive FC_PURE_GRANT by some reason in normal flow

Comment (evgeny-leksikov) [reply test/gtest/ucp/test_ucp_sockaddr.cc:76]: then it seems unrelated

File: src/ucp/core/ucp_worker.c (hunk 3 above)
Comment (yosefe) [reply src/ucp/core/ucp_worker.c:3021]: add comment for this?

File: src/ucp/stream/stream_recv.c [hunk 10]
```diff
@@ -533,6 +533,7 @@ ucp_stream_am_handler(void *am_arg, void *am_data, size_t am_length,
 
     ucs_assert(am_length >= sizeof(ucp_stream_am_hdr_t));
 
+    /* drop the date if the endpoint is invalid */
```
Comment (yosefe) [inline src/ucp/stream/stream_recv.c:536]: Drop the data

**Review by yosefe — APPROVED** [review]

**Comment by yosefe** [comment]
@evgeny-leksikov can you pls squash?

**Comment by evgeny-leksikov** [comment]
bot:pipe:retest

**Pull request merged**